api/wasm/rust/src/lib.rs

```diff
@@ -0,0 +1,72 @@
+/// Logger that integrates with host's logging system.
+pub struct Logger;
+
+static LOGGER: Logger = Logger;
+
+impl Logger {
+    pub fn init() -> Result<(), log::SetLoggerError> {
+        log::set_logger(&LOGGER).map(|()| log::set_max_level(log::LevelFilter::Trace))
+    }
+
+    fn proxywasm_loglevel(level: log::Level) -> u32 {
+        match level {
+            log::Level::Trace => 0,
+            log::Level::Debug => 1,
+            log::Level::Info => 2,
+            log::Level::Warn => 3,
+            log::Level::Error => 4,
+        }
+    }
+}
+
+impl log::Log for Logger {
+    fn enabled(&self, _metadata: &log::Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &log::Record) {
+        let level = Logger::proxywasm_loglevel(record.level());
+        let message = record.args().to_string();
+        unsafe {
+            host::proxy_log(level, message.as_ptr(), message.len());
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+/// Always hook into host's logging system.
+#[no_mangle]
+fn __post_instantiate() {
+    Logger::init().unwrap();
+}
+
+/// Allow host to allocate memory.
+#[no_mangle]
+fn _malloc(size: usize) -> *mut u8 {
+    let mut vec: Vec<u8> = Vec::with_capacity(size);
+    unsafe {
+        vec.set_len(size);
+    }
+    let slice = vec.into_boxed_slice();
+    Box::into_raw(slice) as *mut u8
+}
+
+/// Allow host to free memory.
+// TODO(PiotrSikora): make sure ptr is within acceptable range.
```

Comment:
This should be the responsibility of the surrounding code.   Wasm::freeMemory calls wasm_vm_->getMemoryOffset which returns false if the address is out of range.

Reply:
Agreed, I was going back and forth on where it should be handled (I was mostly thinking about making sure that `free()` isn't called for the addresses below `__heap_base`, etc., but it's actually a bit tricky to do from within the WASM code).

I'll drop the `TODO`. Thanks!